### ISSUE TITLE
Declare authenticated pages in blitz.config.js

### ISSUE THREAD

**Issue by JuanM04** [blitz-js/blitz#1918]
### What do you want and why?

A way to declare [authenticated pages](https://canary.blitzjs.com/docs/authorization#secure-your-pages) inside `blitz.config.js`. This would be helpful if an app has a big chunk of authenticated pages.

### Possible implementation(s)

A possible implementation could be this:

```js
module.exports = {
  securePages: [
    { path: '/app', authenticate: { redirectTo: '/login' } },
    { path: '/app/(.*)', authenticate: { redirectTo: '/login' } },
    { path: '/login', redirectAuthenticatedTo: '/app' },
    { path: '/register', redirectAuthenticatedTo: '/app' },
  ]
}
```

Each item of `securePages` must have a `path` (following [path-to-regexp](https://www.npmjs.com/package/path-to-regexp), the same library Next.js uses for `headers` and `rewrites`) and either an `authenticate` or `redirectAuthenticatedTo` key.

Maybe, this could overwritten by `Page.authenticate`

**Comment by flybayer** [comment]
Yes, great idea!! We can implement this with a babel plugin that reads the config and adds `Page.x` as needed.

I think maybe `routesConfig` is a better key which allows other configs other than auth.

```ts
module.exports = {
  routesConfig: [
    { path: '/app', authenticate: { redirectTo: '/login' } },
    { path: '/app/(.*)', authenticate: { redirectTo: '/login' } },
    { path: '/login', redirectAuthenticatedTo: '/app' },
    { path: '/register', redirectAuthenticatedTo: '/app' },
  ]
}
```

**Comment by kivi** [comment]
authenticate and redirectAuthenticatedTo as predefined conditions.
And and you could add customFunction: checkSomething(req, {redirectTo: '/', reject: response404()})

**Comment by flybayer** [comment]
> And and you could add customFunction: checkSomething(req, {redirectTo: '/', reject: response404()})

We could do this, but this would add `getServerSideProps` to every page, de-opting out of static pre-rendering. But this could certainly be desirable in various cases.

**Comment by flybayer** [comment]
So I just had another idea.

What if instead of this route config, we add first class `Layout` support that lets you do `Layout.authenticate = true` so that all pages with that layout are authenticated. And you could disable per page with `Page.authenticate = false`.

**Comment by benbender** [comment]
If this is extended anyways, it would be super nice to support Roles too. So that I could do 

```js
Layout.authenticate = ["Admin"]
```

Instead of

```js
Layout.authenticate = true
```

Role based auth-checks are possible nearly everywhere afaik in the framework and the exception is quite counter-intuitive...

**Comment by flybayer** [comment]
@benbender I 100% agree. The implementation is going to be complex which is why I didn't add it on the first pass
